wp.data: Split store implementation out from registry.

coderkevin opened the pull request:
## Description
The goal of this PR is to split out the implementation of `@wordpress/data` store implementations to be separate from the registry, opening up the registry for alternative data implementations that can all share a common interface. The ethos behind this is to allow `@worpdress/data` to be a universal interface for data within Gutenberg and WordPress core, regardless of the implementation of the data system backing it. The benefit of doing this is so existing data systems can be integrated with `@wordpress/data` for interoperation or transition purposes.

The common interface for data to be registered will be:

```
registry.registerGenericStore( key, { getSelectors, getActions, subscribe } );
```

* `getSelectors()` will return an object of pre-bound named selector functions.
* `getActions()` will return an object of pre-bound functions.
* `subscribe()` will behave as a redux store subscribe.

Note: This deprecates the following functions on the registry in favor of using `registerStore` and `registerGenericStore`.

* `registerReducer` (use `registerStore` instead)
* `registerSelectors` (use `registerStore` instead)
* `registerActions` (use `registerStore` instead)
* `registerResolvers` (use `registerStore` instead)
* `use` (implement using `registerGenericStore` instead)

## How has this been tested?

I purposely left existing tests alone in this PR to ensure that they all run the exact same way as before.
I also added a few tests to verify expectations of `registerGenericStore` parameter types. No difference in operation should be perceptible.

## Types of changes

This is a refactor of the registry, to remove the implicit store implementation from it.
All code and tests should still behave the same for `@wordpress/data` and the Gutenberg editor.

After the initial interface here is settled, I plan on creating at least one example PR in another repo to validate the operation of using another implementation via `registerGenericStore`

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

coderkevin committed:
data: Start split of registry with reducer

This splits out the implementation of `registerReducer` into a separate
file, with the intention of continuing the rest of the store
implementation. This will make room for other possible implementations
with a common interface.

coderkevin committed:
data: Move registerSelectors and registerActions

This moves the implementation of registerSelectors and registerActions
to the namespace-store implementation file.

notnownikki commented:
How do I use resolvers with this new registration method?

youknowriad commented:
@notnownikki It shouldn't change a lot once this is ready. This is basically adding a new `register` function but the existing one will remain, it will just use this newly introduced function under the hood.

notnownikki commented:
Oh thank goodness for that, resolvers are sweet sweet magic :)

coderkevin committed:
data: Untangle registerResolvers

This untangles registerResolvers, puts more code into the
namespace-store, and organizes the functions upon which it depends.

coderkevin committed:
data: Centralize namespace registration

This commit centralizes the logic for namespace registration, in order
to support `registerStore` better.
`register[Reducer|Actions|Selectors|Resolvers]` now are viewed as
incomplete calls to `registerStore`

coderkevin committed:
data: Add registerGenericStore

This adds the registerGenericStore function and removes the `namespaces`
object in favor of a more generic `stores` object.

coderkevin committed:
data: Use store.subscribe() function

coderkevin commented:
@youknowriad, @aduth, @gziolo, and others, I think this PR is ready for reviewing. There are other things I'd like to do, like deprecating some functions and the plugin mechanism, then making store configs immutable when added (except to store state, of course), which would simplify things further, but that will take more work to untangle.

I'm satisfied with this as a first step into separating out the interfaces from the implementation, and it should allow alternative stores to be implemented.

coderkevin commented on packages/data/src/registry.js:
```diff
@@ -278,19 +139,108 @@ export function createRegistry( storeConfigs = {} ) {
+			throw new TypeError( 'config.getActions must be a function' );
+		}
+		if ( ! config.subscribe ) {
+			throw new TypeError( 'config.subscribe must be a function' );
```
Yes, it should. Will change, thanks!

coderkevin replied:
Fixed in 3525ee3

aduth commented on packages/data/src/registry.js:
```diff
@@ -278,19 +139,108 @@ export function createRegistry( storeConfigs = {} ) {
+			throw new TypeError( 'config.subscribe must be a function' );
+		}
+		stores[ key ] = config;
+		config.subscribe( globalListener );
```
Having the custom store be responsible for tracking its own subscribers seems like a non-trivial task. I wonder if this should be inverted, where the custom store invokes a callback provided by the data module to be called when its own internal state is known to have been updated (callback as return value of `registerGenericStore`?).

coderkevin replied:
That's an interesting idea. I can try it out.

coderkevin replied:
So, I tried this out a bit, and I'm not sure that I like the idea of just assuming that the code which called `registerGenericStore` hooks up the event handler correctly. I feel that the `subscribe` method is a better contract. 

I did add a test for this, and made the simplest subscribe function I could. Since we can assume only one listener for this application, and no need to unsubscribe, it can be very simple:
```
	let listener = () => {};
	const storeChanged = () => {
		listener();
	};
	subscribe = ( newListener ) => {
		listener = newListener;
	};
```

aduth reviewed (commented):
>`getSelectors()` will return an object of pre-bound named selector functions.
>`getActions()` will return an object of pre-bound functions.

To one who is creating a non-Redux-based store, what does "pre-bound" mean to them?

To this and my earlier comment, I'm curious what a minimal example of a custom store might look like. This seems like a good thing to include in `README.md` anyways.

gziolo commented:
I'm moving it to 4.2 as we decided that 4.1 should be UI freeze focused. 4.2 is planned to be focused on API freeze and this PR fits perfectly to that description :)

coderkevin committed:
data: Make registerGenericStore public + add tests

This makes `registry.registerGenericStore` accessible from the result of
`createRegistry`, and adds tests for the functionality of it.

coderkevin committed:
data: Add registerGenericStore to readme

This adds information about `registerGenericStore` to the readme with
two examples.

coderkevin commented:
@youknowriad, @aduth, @gziolo, and others, I have updated the README and added tests for `registerGenericStore`. Feel free to take a look at both of those for example code of how this will be used. Please review and let me know if you have further questions. Thanks!

coderkevin committed:
data: Export registerGenericStore from index

This adds an export for `registerGenericStore` for the defaultRegistry.
While the defaultRegistry may be going away in the future, it's here for
now, so this should be there until it changes.

psealock commented on packages/data/README.md:
```diff
@@ -286,6 +286,93 @@ const SaleButton = withDispatch( ( dispatch, ownProps ) => {
+} );
+
+const genericStore = {
+	getSelectors(): {
```
You can remove the colon. Not valid js, right? Also in examples below

coderkevin replied:
Yes, thanks again.

ocean90 commented on packages/data/README.md:
````diff
@@ -286,6 +286,93 @@ const SaleButton = withDispatch( ( dispatch, ownProps ) => {
+registry.registerGenericStore( 'existing-app', genericStore );
+```
+
+It is also possible to impelement a completely custom store from scratch:
````
Typo: implement

coderkevin replied:
Thanks!

coderkevin commented on packages/data/README.md:
```diff
@@ -286,6 +286,93 @@ const SaleButton = withDispatch( ( dispatch, ownProps ) => {
+	};
+}
+
+registry.registerGenericStore( 'custom-data', customStore );
```
Yes, thanks! I changed the above but missed this line.

coderkevin committed:
data: Fixes to readme

youknowriad reviewed (commented):
Do you have time to rebase this PR. As we simplified a lot of things in the data module in master.

Also, should we add deprecations for `register*` functions and `use`?

youknowriad commented:
Moving out of 4.2 for now as it doesn't seem close.

coderkevin commented:
Sorry @youknowriad, I've been out ill for a couple days. I'll get this rebased by tomorrow.

coderkevin committed:
data: Deprecate registerX and use functions

This deprecates the following functions from the registry:

- registerReducer
- registerSelectors
- registerActions
- registerResolvers
- use

registerStore should be used instead of the above register functions,
and registerGenericStore should be used instead of plugins.

coderkevin commented:
I have rebased on top of current master for today and added deprecation calls for the following:

* 'registerReducer'
* 'registerSelectors'
* 'registerActions'
* 'registerResolvers'
* 'use'

The register functions above should use `registerStore` instead. And code that calls `use` should be able to call `registerGenericStore` instead to implement what it needs.

coderkevin commented:
@youknowriad, @aduth, @gziolo, and others, may I humbly ask that we try to get this PR merged before any other large changes happen to the `registry.js` file? It took the better part of a day to get through the rebase on this and I'd prefer not to do that again.

coderkevin committed:
data: docs/changelog updates for deprecations

This documents the deprecated.md and CHANGELOG.md files with
information about th wp.data registry deprecations.
It also sets the correct version for the deprecated calls to the next
minor plugin version.

coderkevin committed:
data: un-deprecate `registry.use`

This removes the deprecation of `use` because it's internally
referenced and breaks unit tests. The deprecation of `use` should be
done in a future PR that changes the behavior to not call `use`.

youknowriad commented on packages/data/README.md:
````diff
@@ -286,6 +286,93 @@ const SaleButton = withDispatch( ( dispatch, ownProps ) => {
+}
+
+registry.registerGenericStore( 'custom-data', createCustomStore() );
+```
````
❤️

youknowriad commented on packages/data/package.json:
```diff
@@ -23,6 +23,7 @@
 	"dependencies": {
 		"@babel/runtime": "^7.0.0",
 		"@wordpress/compose": "file:../compose",
+		"@wordpress/deprecated": "file:../deprecated",
```
This means we should probably add `wp-deprecated` as a dependency of the `wp-data` script in `client-assets.php`

coderkevin replied:
Added in d0c6dba

youknowriad commented on packages/data/src/plugins/persistence/test/index.js:
```diff
@@ -21,6 +21,7 @@ describe( 'persistence', () => {
 
 		// Since the exposed `registerStore` is a proxying function, mimic
 		// intercept of original call by adding an initial plugin.
+		// TODO: Remove the `use` function in favor of `registerGenericStore`
```
Let's follow up quickly with a PR to deprecate `use` and make the persistence and controls parts of the "namespace stores". I'd like if possible to include all these deprecations in the same version.

coderkevin replied:
Is this something you're looking for my help with or do you have someone in mind to do this? Just asking for time allocation purposes.

youknowriad replied:
I'm not particularly looking for help. I think it's important because otherwise, I feel this refactoring is "incomplete". Anyone could do it thought.

coderkevin replied:
I created issue #11449 for this. I've got a couple other things to attend to first, but I'll see if I can get to this afterwards.

youknowriad reviewed (approved):
Looks good to me overall, I'd appreciate a second opinion cc @aduth

coderkevin committed:
data: Add deprected as script dependency to data

This adds the `wp-deprecated` script as a js script dependency to
`wp-data`.

aduth reviewed (approved):
This is great work 👍

coderkevin commented:
Thanks @youknowriad and @aduth for the reviews and support! 🎉
